neps/nep-0536.md

```diff
@@ -0,0 +1,104 @@
+---
+NEP: 536
+Title: Reduce the number of gas refunds
+Authors: Evgeny Kuzyakov <ek@fastnear.com>, Bowen Wang <bowen@near.org>
+Status: New
+DiscussionsTo: https://github.com/near/NEPs/pull/536
+Type: Protocol
+Version: 1.0.0
+Created: 2024-03-12
+LastUpdated: 2023-03-12
+---
+
+## Summary
+
+[Gas refund](https://docs.near.org/concepts/basics/transactions/gas#attach-extra-gas-get-refunded) is a mechanism that allows users to get refunded for gas that is not used during the execution of a smart contract. Due to [pessimistic gas pricing](https://docs.near.org/concepts/basics/transactions/gas-advanced#pessimistic-gas-price-inflation), however, even transactions that do not involve function calls generate refunds because users need to pay at a high price and get refunded the difference. Gas refunds lead to nontrivial overhead in runtime and other places, which hurts the performance of the protocol. This proposal aims to reduce the number of gas refunds and prepare for future changes that completely remove gas refunds.
+
+## Motivation
+
+Refund receipts create nontrivial overhead: they need to be merklized and sent across shards. In addition, the processing of refund receipts requires additional storage reads and writes, which is not optimal for the performance of the protocol. In addition, when there is congestion, refund receipts may be delayed during execution. Whenever this happens, it requires two additional storage writes to store a gas refund receipt and two additional reads and writes when they are later processed, which incurs a significant overhead. To optimize the performance of the protocol under congestion, it is imperative that we reduce the number of refund receipts.
+
+## Specification
+
+Pessimistic gas pricing is removed as a part of this change. This means that transactions that do not involve function calls will not generate gas refund receipts as a result. For function calls, this proposal introduces cost of refund to be
+```
+REFUND_FIXED_COST = action_receipt_creation + action_transfer + action_add_function_call_key
+refund_cost = max(REFUND_FIXED_COST, 0.05 * gas_refund);
```

Comment:
I think it would be nice to justify this 5% number. How is it chosen instead of 6% or 4%? What network data is used to make this decision?

Reply:
It is an arbitrary number. The goal here is to start penalizing users for attaching too much gas. The percentage should be large enough so that someone attaching 100Tgas or more is penalized for doing so but also small enough so that a callback with a fixed amount of gas reserved are not heavily impacted because the contract is potentially already locked. I think a smaller percentage would work well too.

Reply:
> small enough so that a callback with a fixed amount of gas reserved are not heavily impacted

Shouldn't fixed gas callbacks be unaffected by this change? This change effectively makes refunds smaller, but there is no refund for gas reserved for promises until after those receipts have already executed.

Reply:
The callbacks themselves are unaffected, but the amount of gas refund would be